core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -632,6 +711,524 @@ private boolean checkForFullMatch(InFlightBatch inFlightBatch, long firstOffsetT
+             batches from cachedState is as follows:
+             a) Only full batches can be removed from the cachedState, For example if there is batch (0-99)
+             and 0-49 records are acknowledged (ACCEPT or REJECT), the first 50 records will not be removed
+             from the cachedState. Instead, the startOffset will be moved to 50, but the batch will only
+             be removed once all the messages (0-99) are acknowledged (ACCEPT or REJECT).
+            */
+
+            // Since only a subMap will be removed, we need to find the first and last keys of that subMap
+            long firstKeyToRemove = cachedState.firstKey();
+            long lastKeyToRemove;
+            NavigableMap.Entry<Long, InFlightBatch> entry = cachedState.floorEntry(lastOffsetAcknowledged);
+            if (lastOffsetAcknowledged == entry.getValue().lastOffset()) {
+                startOffset = cachedState.higherKey(lastOffsetAcknowledged);
+                lastKeyToRemove = entry.getKey();
+            } else {
+                startOffset = lastOffsetAcknowledged + 1;
+                if (entry.getKey().equals(cachedState.firstKey())) {
+                    // If the first batch in cachedState has some records yet to be acknowledged,
+                    // then nothing should be removed from cachedState
+                    lastKeyToRemove = -1;
+                } else {
+                    lastKeyToRemove = cachedState.lowerKey(entry.getKey());
+                }
+            }
+
+            if (lastKeyToRemove != -1) {
+                NavigableMap<Long, InFlightBatch> subMap = cachedState.subMap(firstKeyToRemove, true, lastKeyToRemove, true);
+                for (Long key : subMap.keySet()) {
+                    cachedState.remove(key);
+                }
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    private boolean canMoveStartOffset() {
+        // The Share Partition Start Offset may be moved after acknowledgement request is complete.
+        // The following conditions need to be met to move the startOffset:
+        // 1. When the cachedState is not empty.
+        // 2. When the acknowledgement type for the records is either ACCEPT or REJECT.
+        // 3. When all the previous records have been acknowledged (ACCEPT or REJECT).
+        if (cachedState.isEmpty()) {
+            return false;
+        }
+
+        NavigableMap.Entry<Long, InFlightBatch> entry = cachedState.floorEntry(startOffset);
+        if (entry == null) {
+            log.error("The start offset: {} is not found in the cached state for share partition: {}-{}."
+                + " Cannot move the start offset.", startOffset, groupId, topicIdPartition);
+            return false;
+        }
+        RecordState startOffsetState = entry.getValue().offsetState == null ?
+            entry.getValue().batchState() :
+            entry.getValue().offsetState().get(startOffset).state();
```

Comment:
For compacted topics there could be 2 possibilities:
1) The gap exists within the batch i.e. batch fetch from 10 - 20 and gap exists at 15 offset
2) The gap exists outside of the fetched batches i.e. fetched batches are 10-20, 23-35, the gap exists at 21 and 22.

For 1 the gap is identified by the client where offset is moved to Archived state and for 2 the gaps are never filled up in in-memory skip list.

However as the state is fetched directly on startOffset hence there might be a case where we can have NPE. I have created a task for myself to work on it: https://issues.apache.org/jira/browse/KAFKA-17351

Reply:
Just to be clear. The case that I was referring to is the gap between baseOffset and firstOffset. Consider a batch with baseOffset 100 and firstOffset of 105. In the following code, InFlightBatch.offsetState will start at offset 105.

```
        private void maybeInitializeOffsetStateUpdate() {
            if (offsetState == null) {
                offsetState = new ConcurrentSkipListMap<>();
                // The offset state map is not initialized hence initialize the state of the offsets
                // from the first offset to the last offset. Mark the batch inflightState to null as
                // the state of the records is maintained in the offset state map now.
                for (long offset = this.firstOffset; offset <= this.lastOffset; offset++) {
``` 

We set startOffset from either baseOffset of a batch or lastOffset of a batch + 1. So startOffset could be 100 or less than 100 (if previous batch has more than one offset gap at the end). In either case, we will hit `NullPointerException` when calling `offsetState().get(startOffset)`.

Reply:
I might be mistaken but still trying to understand how this could happen. The base offset will always be the first offset of the batch, the situation where baseOffset and firstOffset within a batch should not exist.

Reply:
This is the tricky part with compacted topics. The following is the code in `MemoryRecords` for generating a new batch for retained records after cleaning. The code uses the baseOffset from the original batch as the baseOffset, which can be different from the offset of the first retained record.

```
    private static MemoryRecordsBuilder buildRetainedRecordsInto(RecordBatch originalBatch,
                                                                 List<Record> retainedRecords,
                                                                 ByteBufferOutputStream bufferOutputStream,
                                                                 final long deleteHorizonMs) {
        byte magic = originalBatch.magic();
        Compression compression = Compression.of(originalBatch.compressionType()).build();
        TimestampType timestampType = originalBatch.timestampType();
        long logAppendTime = timestampType == TimestampType.LOG_APPEND_TIME ?
                originalBatch.maxTimestamp() : RecordBatch.NO_TIMESTAMP;
        long baseOffset = magic >= RecordBatch.MAGIC_VALUE_V2 ?
                originalBatch.baseOffset() : retainedRecords.get(0).offset();

        MemoryRecordsBuilder builder = new MemoryRecordsBuilder(bufferOutputStream, magic,
                compression, timestampType, baseOffset, logAppendTime, originalBatch.producerId(),
                originalBatch.producerEpoch(), originalBatch.baseSequence(), originalBatch.isTransactional(),
                originalBatch.isControlBatch(), originalBatch.partitionLeaderEpoch(), bufferOutputStream.limit(), deleteHorizonMs);

        for (Record record : retainedRecords)
            builder.append(record);

```